src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+        OMPSingleDirective (If ancestor OMPSingle has nowait=False)
+        OMPParallelDirective
+
+        The forward dependency is never a child of taskloop, and must have
+        abs_position > taskloop.abs_position
+
+        :param taskloop: the taskloop node to search for the
+                         forward_dependence.
+        :type taskloop: :py:class:`psyclone.psyir.nodes.OMPTaskloopDirective
+        :param root: the tree in which to search for the forward_dependence.
+        :type root: :py:class:`psyclone.f2pygen.DoGen`
+
+        :returns: the forward_dependence of taskloop.
+        :rtype: :py:class:`psyclone.f2pygen.DoGen`
+        '''
+        # We only look for specific types
+        node_list = root.walk((Loop, OMPDoDirective, OMPTaskloopDirective,
+                               OMPTaskwaitDirective, OMPSingleDirective,
+                               OMPParallelDirective))
+        taskloop_vars = VariablesAccessInfo()
+        for child in taskloop.walk(nodes.Node):
+            if child is not taskloop and not isinstance(child,
+                                                        (Schedule, Loop)):
+                taskloop_vars.merge(VariablesAccessInfo(child))
+        taskloop_signatures = taskloop_vars.all_signatures
+        # Find our parent serial region if it has a barrier
+        parent_single = taskloop.ancestor(OMPSingleDirective)
+        if parent_single is not None and parent_single.nowait:
+            parent_single = None
+        # Find our parent parallel region
+        parent_parallel = taskloop.ancestor(OMPParallelDirective)
+        for node in node_list:
+            if node.abs_position <= taskloop.abs_position:
+                continue
+            # Ignore our children
+            anc = node.ancestor(OMPTaskloopDirective)
+            if anc is taskloop:
+                continue
+            node_vars = None
+            # Special case:
+            # 1. If we have a different parent parallel node, then our parent
+            # parallel node is our dependency
+            if node.ancestor(OMPParallelDirective) is not parent_parallel:
+                return parent_parallel
+
+            if (isinstance(node, OMPTaskwaitDirective) and
+                    (taskloop.ancestor(OMPSingleDirective) is
+                     node.ancestor(OMPSingleDirective)) and
+                    taskloop.ancestor(OMPSingleDirective) is not None and
+                    taskloop.ancestor(OMPSingleDirective).nowait is False):
+                # If we find a taskwait barrier inside the same
+                # OMPSingleDirective and that OMPSingleDirective has a no
+                # nowait clause, then it acts as a barrier for dependencies
+                # as well
+                node_vars = taskloop_vars
+            elif not isinstance(node, (OMPSerialDirective,
+                                       OMPParallelDirective)):
+                # For all our other node types that cannot be our parents,
```

Comment:
I don't understand this comment. Perhaps this relates to what the expected relationship between `taskloop` and `root` is? Please could you clarify and maybe add a validation step to check that things are as you expect upon entry to this routine?

Reply:
Actually it is never a parent of OMPTaskloop by this check:
```
if node.abs_position <= taskloop.abs_position:
                  continue
```

Reply:
This has been rewritten but hopefully clears up a lot of the confusing behaviour.